### PR TITLE
Allow module, submodule, and program names to be used as local identi…

### PR DESCRIPTION
**Pull request by klausler**
…fiers.  Check for USE of m in m.

### CODE REVIEW

File: lib/semantics/resolve-names.cc [hunk 1]
```diff
@@ -2086,14 +2090,17 @@ void ModuleVisitor::BeginModule(const parser::Name &name, bool isSubmodule) {
...
+  Scope *scope{reader.Read(name.source, ancestor)};
   if (!scope) {
     return nullptr;
   }
   if (scope->kind() != Scope::Kind::Module) {
     Say(name, "'%s' is not a module"_err_en_US);
     return nullptr;
   }
+  if (DoesScopeContain(scope, currScope())) {
+    Say(name, "Module '%s' cannot USE itself."_err_en_US);
```
Comment (tskeith) [inline lib/semantics/resolve-names.cc]: Should return `nullptr` here too?

File: lib/semantics/resolve-names.cc [hunk 2]
```diff
@@ -1653,7 +1653,11 @@ void ScopeHandler::PushScope(Scope &scope) {
   if (kind != Scope::Kind::Block) {
     ImplicitRulesVisitor::BeginScope(scope);
   }
-  if (!currScope_->IsDerivedType()) {
+  // The name of a module or submodule cannot be "used" in its scope,
+  // as we read 19.3.1(2), so we allow the name to be used as a local
```
Comment (vdonaldson) [inline lib/semantics/resolve-names.cc]: Did you mean C1406 here?

File: lib/semantics/resolve-names.cc (hunk 1 above)
Comment (vdonaldson) [inline lib/semantics/resolve-names.cc]: Add a C1406 comment here?

Comment (klausler) [reply lib/semantics/resolve-names.cc]: I worried that doing so would elicit needless error messages in the following code.

File: lib/semantics/resolve-names.cc (hunk 2 above)
Comment (klausler) [reply lib/semantics/resolve-names.cc]: No.

File: lib/semantics/resolve-names.cc (hunk 1 above)
Comment (klausler) [reply lib/semantics/resolve-names.cc]: I don't think that C1406 applies.  I'll reference the relevant text.

**Comment by klausler** [comment]
Tim, any remaining concern?

File: lib/semantics/resolve-names.cc (hunk 1 above)
Comment (vdonaldson) [reply lib/semantics/resolve-names.cc]: R1409 use-stmt is USE ...

C1406 (R1409) A scoping unit shall not directly reference an intrinsic module and a nonintrinsic module of the same name.

Seems like this is what the message is checking.  I don't get any valid hits on C1406 in f18 code.

**Comment by tskeith** [comment]
> Tim, any remaining concern?

Nope, just forgot to hit Approve.

File: lib/semantics/resolve-names.cc (hunk 1 above)
Comment (klausler) [reply lib/semantics/resolve-names.cc]: That constraint prevents
```
USE,INTRINSIC::IEEE_ARITHMETIC
USE,NON_INTRINSIC::IEEE_ARITHMETIC`
```

Comment (vdonaldson) [reply lib/semantics/resolve-names.cc]: I see; the conjunction _and_ vs. _or_ makes the difference from what I was reading.  Sorry for the extra cycles.

**Pull request merged**